### PR TITLE
Add support for EP11 token to import EC keys

### PR DESCRIPTION
**Pull request by tuchscherer**
These patches enable the EP11 token to import clear key EC private and public keys as well as secure EC key blobs. 
Test cases for several curves were added to the existing EC test cases.

### CODE REVIEW

**Comment by dodys** [comment]
Hi Ingo, your branch is outdated by 11 commits. 
Although, as we discussed, I am fixing the configure.ac, there is some problem in your patches, that lead to the following errors:
ep11_specific.c: In function 'import_EC_key':
ep11_specific.c:1491:28: error: 'raw2key_wrap_blob' undeclared (first use in this function)
   rc = dll_m_EncryptSingle(raw2key_wrap_blob, raw2key_wrap_blob_l, &mech_w,
                            ^
ep11_specific.c:1491:28: note: each undeclared identifier is reported only once for each function it appears in
ep11_specific.c:1491:47: error: 'raw2key_wrap_blob_l' undeclared (first use in this function)
   rc = dll_m_EncryptSingle(raw2key_wrap_blob, raw2key_wrap_blob_l, &mech_w,
                                               ^
ep11_specific.c:1493:24: error: 'struct _STDLL_TokData_t' has no member named 'target_list'
       (uint64_t)tokdata->target_list);
                        ^
ep11_specific.c:1516:16: error: 'ep11_pin_blob' undeclared (first use in this function)
      NULL, ~0, ep11_pin_blob, ep11_pin_blob_len, &mech_w,
                ^
ep11_specific.c:1516:31: error: 'ep11_pin_blob_len' undeclared (first use in this function)
      NULL, ~0, ep11_pin_blob, ep11_pin_blob_len, &mech_w,
                               ^
ep11_specific.c:1518:23: error: 'struct _STDLL_TokData_t' has no member named 'target_list'
      (uint64_t)tokdata->target_list);
                       ^
This has to do with the changes done by Franzki on 8386f5a.
Please, sync your master branch with ock current master and then do a rebase master on your ep11-ec-keyimport and fix what is needed (doing a interactive rebase of your patches should be enough).

**Comment by tuchscherer** [comment]
Hi Eduardo,
Sorry for the inconvenience. I have re-based this feature branch. It now adapts to the changes from Ingo Franzki. It still needs the linker options for 'lber' to build completely.

**Comment by dodys** [comment]
Thanks Ingo!
I will test it with the changes done in configure.ac.

File: usr/lib/pkcs11/common/key.c [hunk 1]
```diff
@@ -2236,6 +2236,15 @@ ecdsa_priv_validate_attribute( STDLL_TokData_t *tokdata, TEMPLATE *tmpl,
             TRACE_ERROR("%s\n", ock_err(ERR_ATTRIBUTE_READ_ONLY));
             return CKR_ATTRIBUTE_READ_ONLY;
          }
+      case CKA_EC_POINT:
+         if (mode == MODE_CREATE) {
+            p11_attribute_trim( attr );
+            return CKR_OK;
+         }
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: fix if-else statement, please check coding-style documentation

File: usr/lib/pkcs11/common/asn1.c [hunk 2]
```diff
@@ -571,7 +573,99 @@ ber_decode_SEQUENCE( CK_BYTE  * seq,
...
+
+	if (data_len < 128)
+		len = 1 + 1 + data_len;
+	else if (data_len < 256)
+		len = 1 + (1 + 1) + data_len;
+	else if (data_len < (1 << 16))
+		len = 1 + (1 + 2) + data_len;
+	else if (data_len < (1 << 24))
+		len = 1 + (1 + 3) + data_len;
+	else {
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: if one of the if statements have { }, then all the rest should also have. Please check coding-style.

File: usr/lib/pkcs11/common/asn1.c [hunk 3]
```diff
@@ -571,7 +573,99 @@ ber_decode_SEQUENCE( CK_BYTE  * seq,
...
+		TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+		return CKR_FUNCTION_FAILED;
+	}
+	if (length_only == TRUE) {
+		*str_len = len;
+		return CKR_OK;
+	}
+
+	buf = (CK_BYTE *)malloc( len );
+	if (!buf){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: space needed between ) and {

File: usr/lib/pkcs11/common/asn1.c [hunk 4]
```diff
@@ -571,7 +573,99 @@ ber_decode_SEQUENCE( CK_BYTE  * seq,
...
+		buf[4] = (data_len      ) & 0xFF;
+
+		memcpy( &buf[5], data, data_len );
+
+		*str_len = len;
+		*str = buf;
+		return CKR_OK;
+	}
+
+	free( buf );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: do not include space, write as: free(buf);

File: usr/lib/pkcs11/common/asn1.c [hunk 5]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
 
    return rc;
 }
+
+
+//
+//
+CK_RV
+der_encode_ECPrivateKey( CK_BBOOL    length_only,
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove space between ( and first argument

File: usr/lib/pkcs11/common/asn1.c [hunk 6]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+	BerValue   *val;
+	CK_RV      rc = 0;
+
+	/* Calculate BER encoding length
+	 * Inner SEQUENCE of
+	 * Integer (version), OCTET STRING (private key)
+	 * and CHOICE [1] BIT STRING (public key)
+	 */
+	// version
+	rc |= ber_encode_INTEGER( TRUE, NULL, &len, NULL, sizeof(version) );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 7]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+
+	// private key octet
+	if (opaque != NULL) {
+		rc |= ber_encode_OCTET_STRING(TRUE, NULL, &len, NULL, opaque->ulValueLen);
+		offset += len;
+	} else {
+		rc |= ber_encode_OCTET_STRING(TRUE, NULL, &len, NULL, point->ulValueLen);
+		offset += len;
+	}
+	if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: space needed between ) and {

File: usr/lib/pkcs11/common/asn1.c [hunk 8]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+			TRACE_DEVEL("ber_encode_SEQUENCE failed\n");
+			return rc;
+		}
+		rc = ber_encode_PrivateKeyInfo( TRUE, NULL, data_len, NULL,
+						der_AlgIdECLen, NULL, len );
+		if (rc != CKR_OK){
+			TRACE_DEVEL("ber_encode_PrivateKeyInfo failed\n");
+			return rc;
+		}
+	return rc;
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: it is not aligned to the rest of the code

File: usr/lib/pkcs11/common/asn1.c [hunk 9]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+		rc = ber_flatten(ber, &val);
+
+		ber_encode_CHOICE(TRUE, 1, &buf2, &len, val->bv_val, val->bv_len);
+		offset += len;
+		ber_free(ber, 1);
+	}
+
+	if (length_only == TRUE) {
+		rc = ber_encode_SEQUENCE( TRUE, NULL, &len, NULL, offset );
+		if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: space needed between ) and {

File: usr/lib/pkcs11/common/asn1.c [hunk 10]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+			TRACE_DEVEL("ber_encode_SEQUENCE failed\n");
+			return rc;
+		}
+		rc = ber_encode_PrivateKeyInfo( TRUE, NULL, data_len, NULL,
+						der_AlgIdECLen, NULL, len );
+		if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: same as above .. I will not mark the rest of the code where it happens

File: usr/lib/pkcs11/common/asn1.c [hunk 11]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+	/* Now starting with the real data */
+	buf = (CK_BYTE *)malloc(offset);
+	if (!buf){
+		TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
+		return CKR_HOST_MEMORY;
+	}
+	offset = 0;
+	rc = 0;
+
+	rc = ber_encode_INTEGER( FALSE, &buf2, &len, version, sizeof(version) );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 12]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+		goto error;
+	}
+	memcpy( buf+offset, buf2, len );
+	offset += len;
+	free( buf2 );
+	buf2 = NULL;
+
+	if (opaque != NULL) {
+	// the CKA_IBM_OPAQUE attrib
+		rc = ber_encode_OCTET_STRING( FALSE, &buf2, &len, (CK_BYTE *)opaque + sizeof(CK_ATTRIBUTE), opaque->ulValueLen );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

Comment (dodys) [reply usr/lib/pkcs11/common/asn1.c]: over the limit of 80 columns

File: usr/lib/pkcs11/common/asn1.c [hunk 13]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+		if (rc != CKR_OK){
+			TRACE_DEVEL("ber_encode_OCTET_STRING failed\n");
+			goto error;
+		}
+		memcpy( buf+offset, buf2, len );
+		offset += len;
+		free( buf2 );
+		buf2 = NULL;
+	} else {
+		rc = ber_encode_OCTET_STRING( FALSE, &buf2, &len, (CK_BYTE *)point  + sizeof(CK_ATTRIBUTE),  point->ulValueLen );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: same as above

File: usr/lib/pkcs11/common/h_extern.h [hunk 14]
```diff
@@ -2736,6 +2740,13 @@ CK_RV    ber_decode_RSAPrivateKey( CK_BYTE     * data,
                                    CK_ATTRIBUTE ** opaque,
 				   CK_BBOOL	   isopaque );
 
+CK_RV    der_encode_ECPrivateKey( CK_BBOOL     length_only,
```
Comment (dodys) [inline usr/lib/pkcs11/common/h_extern.h]: remove space

File: usr/lib/pkcs11/common/h_extern.h [hunk 15]
```diff
@@ -2736,6 +2740,13 @@ CK_RV    ber_decode_RSAPrivateKey( CK_BYTE     * data,
...
+                                  CK_BYTE   ** data,
+                                  CK_ULONG   * data_len,
+                                  CK_ATTRIBUTE * params,
+                                  CK_ATTRIBUTE * point,
+                                  CK_ATTRIBUTE * opaque,
+                                  CK_ATTRIBUTE * pubkey );
```
Comment (dodys) [inline usr/lib/pkcs11/common/h_extern.h]: remove space

File: usr/lib/pkcs11/common/asn1.c [hunk 16]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+	der_AlgIdEC[1] = der_AlgIdEC[1] + params->ulValueLen; // adjust length field
+
+	rc = ber_encode_PrivateKeyInfo( FALSE, data, data_len, der_AlgIdEC,
+					der_AlgIdECLen, buf2, len );
+	if (rc != CKR_OK) {
+		TRACE_ERROR("ber_encode_PrivateKeyInfo failed\n");
+	}
+
+error:
+	if (buf2) free( buf2 );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 17]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+	if (buf)  free( buf );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 18]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+
+		ber_encode_CHOICE(FALSE, 1, &buf2, &len, val->bv_val, val->bv_len);
+		memcpy(buf + offset, buf2, len);
+		offset += len;
+		free( buf2 );
+		buf2 = NULL;
+		ber_free(ber, 1);
+	}
+
+	rc = ber_encode_SEQUENCE( FALSE, &buf2, &len, buf, offset );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 19]
```diff
@@ -1591,3 +1685,152 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+	/* generate optional bit-string of public key */
+	if (pubkey && pubkey->pValue) {
+		ber = ber_alloc_t(LBER_USE_DER);
+		rc  = ber_put_bitstring(ber, pubkey->pValue, pubkey->ulValueLen*8, 0x03);
+		rc  = ber_flatten(ber, &val);
+
+		ber_encode_CHOICE(FALSE, 1, &buf2, &len, val->bv_val, val->bv_len);
+		memcpy(buf + offset, buf2, len);
+		offset += len;
+		free( buf2 );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/h_extern.h [hunk 20]
```diff
@@ -2512,6 +2514,8 @@ CK_RV     ecdsa_priv_check_required_attributes( TEMPLATE *tmpl, CK_ULONG mode );
 CK_RV     ecdsa_priv_set_default_attributes( TEMPLATE *tmpl, CK_ULONG mode );
 CK_RV     ecdsa_priv_validate_attribute( STDLL_TokData_t *tokdata, TEMPLATE *tmpl,
 					 CK_ATTRIBUTE *attr, CK_ULONG mode );
+CK_RV     ecdsa_priv_wrap_get_data( TEMPLATE *tmpl, CK_BBOOL length_only,
```
Comment (dodys) [inline usr/lib/pkcs11/common/h_extern.h]: remove spaces

File: usr/lib/pkcs11/common/key.c [hunk 21]
```diff
@@ -1576,6 +1576,72 @@ rsa_priv_unwrap( TEMPLATE *tmpl,
...
+//
+// ECPrivateKey ::= SEQUENCE {
+//   version Version
+//   privateKey OCTET STRING
+//   parameters [0] ECParameters (OPTIONAL)
+//   publicKey  [1] BIT STRING (OPTIONAL)
+// }
+//
+CK_RV
+ecdsa_priv_wrap_get_data( TEMPLATE  *tmpl,
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: remove space

File: usr/lib/pkcs11/common/key.c [hunk 22]
```diff
@@ -1576,6 +1576,72 @@ rsa_priv_unwrap( TEMPLATE *tmpl,
...
+			  CK_BBOOL   length_only,
+			  CK_BYTE  **data,
+			  CK_ULONG  *data_len )
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: remove space

File: usr/lib/pkcs11/common/key.c [hunk 23]
```diff
@@ -1576,6 +1576,72 @@ rsa_priv_unwrap( TEMPLATE *tmpl,
...
+	CK_ATTRIBUTE *params = NULL;
+	CK_ATTRIBUTE *point  = NULL;
+	CK_ATTRIBUTE *opaque = NULL;
+	CK_ATTRIBUTE *pubkey = NULL;
+	CK_RV      rc;
+
+
+	// compute the total length of the BER-encoded data
+	//
+	if (template_attribute_find(tmpl, CKA_EC_PARAMS, &params) == FALSE){
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: add space between ) and {

File: usr/lib/pkcs11/common/key.c [hunk 24]
```diff
@@ -1576,6 +1576,72 @@ rsa_priv_unwrap( TEMPLATE *tmpl,
...
+		TRACE_ERROR("Could not find CKA_EC_PARAMS for the key.\n");
+		return CKR_FUNCTION_FAILED;
+	}
+	if (template_attribute_find(tmpl, CKA_VALUE, &point) == FALSE){
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: same as previous

File: usr/lib/pkcs11/common/key.c [hunk 25]
```diff
@@ -1576,6 +1576,72 @@ rsa_priv_unwrap( TEMPLATE *tmpl,
...
+		TRACE_ERROR("Could not find CKA_EC_POINT for the key.\n");
+		return CKR_FUNCTION_FAILED;
+	}
+
+	// CKA_IBM_OPAQUE is used for secure key, if it is not available, then
+	// assume using clear key and get rest of attributes required for clear key.
+
+	if (template_attribute_find(tmpl, CKA_IBM_OPAQUE, &opaque) == FALSE){
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: same as previous

File: usr/lib/pkcs11/common/key.c [hunk 26]
```diff
@@ -1576,6 +1576,72 @@ rsa_priv_unwrap( TEMPLATE *tmpl,
...
+		if (template_attribute_find(tmpl, CKA_VALUE, &point) == FALSE){
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: same as previous

File: usr/lib/pkcs11/common/key.c [hunk 27]
```diff
@@ -1576,6 +1576,72 @@ rsa_priv_unwrap( TEMPLATE *tmpl,
...
+			return CKR_FUNCTION_FAILED;
+		}
+	}
+
+	/* check if optional public-key part was defined */
+	template_attribute_find(tmpl, CKA_EC_POINT, &pubkey);
+
+	rc = der_encode_ECPrivateKey(length_only, data, data_len, params,
+				     point, opaque, pubkey);
+	if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/key.c]: same as previous

File: usr/lib/pkcs11/ep11_stdll/ep11_specific.c [hunk 28]
```diff
@@ -408,6 +408,81 @@ ber_encode_RSAPublicKey(CK_BBOOL length_only, CK_BYTE **data, CK_ULONG *data_len
...
+	 *
+	 * SPKI := SEQUENCE {
+	 * 	SEQUENCE {
+	 * 		OID
+	 * 		Parameters
+	 * 	}
+	 * 	BITSTRING public key
+	 * }
+	 */
+	rc = ber_encode_SEQUENCE( TRUE, NULL, &len, NULL, algid_len );
```
Comment (dodys) [inline usr/lib/pkcs11/ep11_stdll/ep11_specific.c]: remove spaces

File: usr/lib/pkcs11/ep11_stdll/ep11_specific.c [hunk 29]
```diff
@@ -1293,6 +1368,181 @@ static CK_RV import_RSA_key(STDLL_TokData_t *tokdata, OBJECT *rsa_key_obj,
...
+
+	/* m_Unwrap builds key blob in the card,
+	 * tell ep11 the attributes the user specified for that key.
+	 */
+	node = ec_key_obj->template->attribute_list;
+	while (node != NULL) {
+		CK_ATTRIBUTE_PTR a = node->data;
+
+		/* ep11 handles this as 'read only' */
+		if (CKA_NEVER_EXTRACTABLE == a->type ||
```
Comment (dodys) [inline usr/lib/pkcs11/ep11_stdll/ep11_specific.c]: add {} to this if

File: usr/lib/pkcs11/common/asn1.c [hunk 30]
```diff
@@ -407,6 +407,74 @@ ber_decode_OCTET_STRING( CK_BYTE  * str,
...
+//
+CK_RV
+ber_decode_BIT_STRING( CK_BYTE  * str,
+		       CK_BYTE ** data,
+		       CK_ULONG * data_len,
+		       CK_ULONG * field_len )
+{
+	CK_ULONG  len, length_octets;
+
+	if (!str){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: space needed

File: usr/lib/pkcs11/common/asn1.c [hunk 31]
```diff
@@ -407,6 +407,74 @@ ber_decode_OCTET_STRING( CK_BYTE  * str,
...
+		TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+		return CKR_FUNCTION_FAILED;
+	}
+	if (str[0] != 0x03){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: same as previous

File: usr/lib/pkcs11/common/asn1.c [hunk 32]
```diff
@@ -1686,6 +1754,40 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
    return rc;
 }
 
+/*
+ * ECC Functions
+ */
+//
+//
+CK_RV
+ecdsa_priv_unwrap_get_data( TEMPLATE *tmpl,
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove space

File: usr/lib/pkcs11/common/asn1.c [hunk 33]
```diff
@@ -1686,6 +1754,40 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+			    CK_BYTE  *data,
+			    CK_ULONG  total_length)
+{
+	CK_ATTRIBUTE *params  = NULL;
+	CK_ATTRIBUTE *point   = NULL;
+	CK_RV      rc;
+
+	rc = der_decode_ECPublicKey( data, total_length, &params, &point);
+
+	if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: add space

File: usr/lib/pkcs11/common/asn1.c [hunk 34]
```diff
@@ -1686,6 +1754,40 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+		TRACE_DEVEL("ber_decode_ECPrivateKey failed\n");
+		return rc;
+	}
+
+	p11_attribute_trim( params );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 35]
```diff
@@ -1686,6 +1754,40 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+	p11_attribute_trim( point );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 36]
```diff
@@ -1686,6 +1754,40 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+
+	rc = template_update_attribute( tmpl, params );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 37]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
 	if (buf)  free( buf );
 	return rc;
 }
+
+//
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: long comments, use /* */

File: usr/lib/pkcs11/common/asn1.c [hunk 38]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
...
+// ECPrivateKey ::= SEQUENCE {
+//   version Version
+//   privateKey OCTET STRING
+//   parameters [0] ECParameters (OPTIONAL)
+//   publicKey  [1] BIT STRING (OPTIONAL)
+// }
+//
+CK_RV
+der_decode_ECPublicKey( CK_BYTE	      * data,
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove space

File: usr/lib/pkcs11/common/asn1.c [hunk 39]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
...
+	CK_BYTE  *param        = NULL;
+	CK_ULONG  param_len;
+	CK_BYTE  *point        = NULL;
+	CK_ULONG  point_len;
+	CK_ULONG  offset = 0;
+	CK_ULONG  field_len, len;
+	CK_RV     rc;
+
+	rc = ber_decode_SEQUENCE( data, &inner_seq, &len, &field_len );
+	if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: add space

File: usr/lib/pkcs11/common/asn1.c [hunk 40]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
...
+		TRACE_DEVEL("ber_decode_PrivateKeyInfo failed\n");
+		return rc;
+	}
+
+	rc = ber_decode_SEQUENCE( inner_seq, &algid, &algid_len, &field_len );
+	if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: add space

File: usr/lib/pkcs11/common/asn1.c [hunk 41]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
...
+		TRACE_DEVEL("ber_decode_SEQUENCE failed\n");
+		return rc;
+	}
+	offset += field_len;
+
+	// Make sure we're dealing with an EC key.
+	// Extract base alg-id of DER encoded EC byte string
+	// and compare against the decoded alg-id from the inner sequence
+	rc = ber_decode_SEQUENCE( der_AlgIdECBase, &algid_ECBase, &len, &field_len );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 42]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
...
+	if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: add space

File: usr/lib/pkcs11/common/asn1.c [hunk 43]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
...
+		TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+		return CKR_FUNCTION_FAILED;
+	}
+
+	// skip the generic EC publ key ID and point to the curve
+	// specific parameter
+	param = algid + algid[1] + 2;
+	param_len = algid_len - algid[1] - 2;
+
+	rc = ber_decode_BIT_STRING( inner_seq + offset, &point, &point_len, &field_len );
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: remove spaces

File: usr/lib/pkcs11/common/asn1.c [hunk 44]
```diff
@@ -1834,3 +1936,108 @@ der_encode_ECPrivateKey( CK_BBOOL    length_only,
...
+	if (rc != CKR_OK){
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: add space

File: testcases/crypto/ec.h [hunk 45]
```diff
@@ -0,0 +1,478 @@
+
```
Comment (dodys) [inline testcases/crypto/ec.h]: New files need to include License Header.

**Comment by ifranzki** [comment]
Ingo,  I found one compiler warning caused by your patch 08:

ec_func.c: In function 'run_TransferECCKeyPairSignVerify':
ec_func.c:738:3: warning: 'wrapped_key' may be used uninitialized in this function [-Wmaybe-uninitialized]
   free(wrapped_key);

The reason is that macros testcase_rw_session and testcase_user_login may go to label testcase_cleanup, but you declare and initialize variable wrapped_key only after using those macros. So when one of these macros go to label testcase, then variable wrapped_key is not initialized, and the "free(wrapped_key);" may have unexpected results. 

Simply decare and initialize this variable before using the macros testcase_rw_session and testcase_user_login. Actually, I would declare ALL variables before any line of real code in a function.

**Comment by tuchscherer** [comment]
Eduardo,
I did the changes according to the style guidelines. Not sure if I have catched all them.
Do you have any check script that can run over patch files?
We have something for the kernel called 'checkpatch' that let you know if there are some style issues.
Anyway, let me know if there are still some issues.

**Comment by dodys** [comment]
Thanks, I will review it.

No, we do not have a script yet, but the idea is to have something like the checkpatch.

File: testcases/common/common.c [hunk 46]
```diff
@@ -350,6 +350,88 @@ CK_RV generate_RSA_PKCS_KeyPair(CK_SESSION_HANDLE session,
...
+                {CKA_KEY_TYPE, &keyType, sizeof(keyType)},
+                {CKA_TOKEN, &true, sizeof(true)},
+                {CKA_PRIVATE, &true, sizeof(true)},
+                {CKA_LABEL, label, sizeof(label)},
+                {CKA_SUBJECT, subject, sizeof(subject)},
+                {CKA_ID, id, sizeof(id)},
+                {CKA_SENSITIVE, &true, sizeof(true)},
+                {CKA_DECRYPT, &true, sizeof(true)},
+                {CKA_SIGN, &true, sizeof(true)},
+		{CKA_EC_PARAMS, params, params_len},
```
Comment (dodys) [inline testcases/common/common.c]: Bad alignment

File: testcases/common/common.c [hunk 47]
```diff
@@ -350,6 +350,88 @@ CK_RV generate_RSA_PKCS_KeyPair(CK_SESSION_HANDLE session,
...
+		{CKA_EC_POINT, pubkey, pubkey_len},
```
Comment (dodys) [inline testcases/common/common.c]: Bad alignment

File: testcases/common/common.c [hunk 48]
```diff
@@ -350,6 +350,88 @@ CK_RV generate_RSA_PKCS_KeyPair(CK_SESSION_HANDLE session,
...
+		{CKA_VALUE, privatekey, privatekey_len}
```
Comment (dodys) [inline testcases/common/common.c]: Bad alignment

File: testcases/common/common.c [hunk 49]
```diff
@@ -350,6 +350,88 @@ CK_RV generate_RSA_PKCS_KeyPair(CK_SESSION_HANDLE session,
...
+        CK_UTF8CHAR     label[] = "An EC public key object";
+        CK_BBOOL        true = TRUE;
+        CK_ATTRIBUTE    template[] = {
+                {CKA_CLASS, &class, sizeof(class)},
+                {CKA_KEY_TYPE, &keyType, sizeof(keyType)},
+                {CKA_TOKEN, &true, sizeof(true)},
+                {CKA_LABEL, label, sizeof(label)},
+                {CKA_ENCRYPT, &true, sizeof(true)},
+                {CKA_VERIFY, &true, sizeof(true)},
+		{CKA_EC_PARAMS, params, params_len},
```
Comment (dodys) [inline testcases/common/common.c]: Bad alignment

File: testcases/common/common.c [hunk 50]
```diff
@@ -350,6 +350,88 @@ CK_RV generate_RSA_PKCS_KeyPair(CK_SESSION_HANDLE session,
...
+		{CKA_EC_POINT, pointq, pointq_len}
```
Comment (dodys) [inline testcases/common/common.c]: Bad alignment

File: testcases/crypto/ec.h (hunk 45 above)
Comment (dodys) [reply testcases/crypto/ec.h]: Still need License Header

File: testcases/crypto/ec_func.c [hunk 51]
```diff
@@ -431,6 +431,315 @@ run_GenerateECCKeyPairSignVerify()
...
+	/* query the slot, check if this mech is supported */
+	rc = funcs->C_GetMechanismInfo(SLOT_ID, mech.mechanism, &mech_info);
+	if (rc != CKR_OK) {
+		if (rc == CKR_MECHANISM_INVALID) {
+			/* no support for EC key gen? skip */
+			testcase_skip("Slot %u doesn't support CKM_ECDSA",
+				(unsigned int) SLOT_ID);
+			goto testcase_cleanup;
+		}
+		else {
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: should be "} else {" 
don't break line

File: testcases/crypto/ec_func.c [hunk 52]
```diff
@@ -431,6 +431,315 @@ run_GenerateECCKeyPairSignVerify()
...
```
Comment (dodys) [inline testcases/crypto/ec_func.c]: should be "} else {"
don't break line

File: usr/lib/pkcs11/common/asn1.c [hunk 53]
```diff
@@ -1591,3 +1753,312 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+	if (rc != CKR_OK) {
+		TRACE_DEVEL("ber_decode_ECPrivateKey failed\n");
+		return rc;
+	}
+
+	p11_attribute_trim(params);
+	p11_attribute_trim(point);
+
+	rc = template_update_attribute(tmpl, params);
+        if (rc != CKR_OK)
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: Bad alignment

File: usr/lib/pkcs11/common/asn1.c [hunk 54]
```diff
@@ -1591,3 +1753,312 @@ ber_decode_DSAPrivateKey( CK_BYTE     * data,
...
+                TRACE_DEVEL("template_update_attribute(CKA_EC_PARAMS) failed\n");
+	rc = template_update_attribute(tmpl, point);
+        if (rc != CKR_OK)
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: Bad alignment

File: usr/lib/pkcs11/ep11_stdll/ep11_specific.c [hunk 55]
```diff
@@ -1293,6 +1375,182 @@ static CK_RV import_RSA_key(STDLL_TokData_t *tokdata, OBJECT *rsa_key_obj,
...
+		    CKA_MODIFIABLE == a->type || CKA_LOCAL == a->type) {
+			;
+		}
+		else {
```
Comment (dodys) [inline usr/lib/pkcs11/ep11_stdll/ep11_specific.c]: should be "} else {"
don't break line

File: usr/lib/pkcs11/common/asn1.c [hunk 56]
```diff
@@ -405,6 +407,74 @@ ber_decode_OCTET_STRING( CK_BYTE  * str,
...
+		      CK_ULONG * field_len)
+{
+	CK_ULONG  len, length_octets;
+
+	if (!str) {
+		TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
+		return CKR_FUNCTION_FAILED;
+	}
+	if (str[0] != 0x03) {
+	TRACE_ERROR("%s\n", ock_err(ERR_FUNCTION_FAILED));
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: Bad alignment

File: usr/lib/pkcs11/common/asn1.c [hunk 57]
```diff
@@ -405,6 +407,74 @@ ber_decode_OCTET_STRING( CK_BYTE  * str,
...
+	return CKR_FUNCTION_FAILED;
```
Comment (dodys) [inline usr/lib/pkcs11/common/asn1.c]: Bad alignment

**Comment by tuchscherer** [comment]
OK, I updated the branch. Sorry for the additional cycles. I think once we have some checkpatch available code checking would be much easier. Thanks again for your review.

**Pull request merged**